In [9]:
# copy-paste this into Colab/VSCode and run
import collections
from typing import Dict, Any, List

# ---------------------------------
# Level 1: Rule Definition
# (The symbolic factors that drive the logic)
# ---------------------------------
FREQ_FACTORS = {"diaria": 20, "semanal": 4}
COMPLEX_FACTORS = {"Baixa": 1, "Média": 3, "Alta": 5}

# ---------------------------------
# Level 2: Core Logic Functions
# (Functions that apply the rules to a single task)
# ---------------------------------
def is_automatable(task: Dict[str, Any]) -> bool:
    """Rule 1: Checando a tarefa dos Candidatos para Automação"""
    return task["frequencia"] in FREQ_FACTORS and task["tipo"] != "Aprovação"

def calculate_automation_score(task: Dict[str, Any]) -> float:
    """Rule 2: Calculando a automação com base nos fatores definidos."""
    freq = FREQ_FACTORS.get(task["frequencia"], 0)
    comp = COMPLEX_FACTORS.get(task["complexidade"], 1)
    time = task["tempo_gasto_min"]
    
    if comp == 0: return 0.0
    
    score = (time * freq) / comp
    return round(score, 2)

# ---------------------------------
# Level 3: Analysis, Orchestration, and Execution
# (The main function that processes the entire dataset and runs the analysis)
# ---------------------------------
def analyze_tasks(task_list: List[Dict[str, Any]]) -> None:
    """Level 3 Analysis: Processando todas as tarefas e relatório."""
    
    # --- Data Processing ---
    automatable_tasks = []
    for task in task_list:
        if is_automatable(task):
            task['score'] = calculate_automation_score(task)
            automatable_tasks.append(task)
            
    # --- Quick Wins Analysis ---
    quick_wins = sorted(automatable_tasks, key=lambda x: x['score'], reverse=True)
    
    # --- Productivity Drain Analysis ---
    productivity_drain = collections.defaultdict(float)
    for task in automatable_tasks:
        monthly_time_min = task["tempo_gasto_min"] * FREQ_FACTORS.get(task["frequencia"], 0)
        productivity_drain[task["responsavel"]] += monthly_time_min
        
    drain_person, drain_hours = "", 0.0
    if productivity_drain:
        drain_person, total_min = max(productivity_drain.items(), key=lambda item: item[1])
        drain_hours = round(total_min / 60, 1)
        
    # --- Immediate Impact Analysis ---
    top_3_impact_min = sum(t["tempo_gasto_min"] * FREQ_FACTORS.get(t["frequencia"], 0) for t in quick_wins[:3])
    top_3_impact_hours = round(top_3_impact_min / 60, 1)

    # --- 4. Generate Report ---
    print("Relatório do agente otimizador de Workflow:")
    print("-------------------------------------------\n")
    print("Plano de automação priorizado (Quick Wins):")
    for i, task in enumerate(quick_wins, 1):
        print(f"{i}. {task['descricao']} ({task['responsavel']}) - Score: {task['score']:.1f}")
    
    print("\nDiagnóstico estratégico:")
    if drain_person:
        print(f">> Dreno de Produtividade Identificado: {drain_person} gasta aproximadamente {drain_hours} horas/mês em tarefas com alto potencial de automação.")
    if top_3_impact_hours > 0:
        print(f">> Impacto Imediato: Automatizar os {min(3, len(quick_wins))} principais \"Quick Wins\" pode liberar cerca de {top_3_impact_hours} horas de trabalho por mês para a equipe.")

    print("\nRecomendação do agente:")
    if quick_wins:
        print(f"- FOCO IMEDIATO: Iniciar a automação da tarefa \"{quick_wins[0]['descricao']}\".")

# --- Main execution block ---
if __name__ == '__main__':
    # NOTE: This block requires the 'tarefas' list (dataset) to be defined.
    tarefas = [
        {"id": 1, "descricao": "Gerar relatório de performance semanal", "responsavel": "Ana", "frequencia": "semanal", "tipo": "Relatório", "complexidade": "Baixa", "tempo_gasto_min": 60},
        {"id": 2, "descricao": "Agendar reuniões de alinhamento", "responsavel": "Beatriz", "frequencia": "diaria", "tipo": "Comunicação", "complexidade": "Baixa", "tempo_gasto_min": 15},
        {"id": 3, "descricao": "Validar e aprovar pull request", "responsavel": "Clara", "frequencia": "pontual", "tipo": "Aprovação", "complexidade": "Alta", "tempo_gasto_min": 20},
        {"id": 4, "descricao": "Exportar e limpar dados de vendas", "responsavel": "Ana", "frequencia": "diaria", "tipo": "Entrada de Dados", "complexidade": "Média", "tempo_gasto_min": 45},
        {"id": 5, "descricao": "Atualizar planilha de orçamento", "responsavel": "Beatriz", "frequencia": "semanal", "tipo": "Entrada de Dados", "complexidade": "Média", "tempo_gasto_min": 90},
        {"id": 6, "descricao": "Enviar lembrete de pendências no Slack", "responsavel": "Clara", "frequencia": "diaria", "tipo": "Comunicação", "complexidade": "Baixa", "tempo_gasto_min": 10},
        {"id": 7, "descricao": "Monitorar logs de erro da aplicação", "responsavel": "Ana", "frequencia": "diaria", "tipo": "Monitoramento", "complexidade": "Média", "tempo_gasto_min": 30}
    ]
    analyze_tasks(tarefas)

Relatório do agente otimizador de Workflow:
-------------------------------------------

Plano de automação priorizado (Quick Wins):
1. Agendar reuniões de alinhamento (Beatriz) - Score: 300.0
2. Exportar e limpar dados de vendas (Ana) - Score: 300.0
3. Gerar relatório de performance semanal (Ana) - Score: 240.0
4. Enviar lembrete de pendências no Slack (Clara) - Score: 200.0
5. Monitorar logs de erro da aplicação (Ana) - Score: 200.0
6. Atualizar planilha de orçamento (Beatriz) - Score: 120.0

Diagnóstico estratégico:
>> Dreno de Produtividade Identificado: Ana gasta aproximadamente 29.0 horas/mês em tarefas com alto potencial de automação.
>> Impacto Imediato: Automatizar os 3 principais "Quick Wins" pode liberar cerca de 24.0 horas de trabalho por mês para a equipe.

Recomendação do agente:
- FOCO IMEDIATO: Iniciar a automação da tarefa "Agendar reuniões de alinhamento".


### Análise de Pontuações Corrigidas
* **1. "Agendar reuniões de alinhamento" (Beatriz) - Pontuação: 300.0**
    * **Tempo:** 15 minutos
    * **Frequência:** `diária` (Fator = 20)
    * **Complexidade:** `Baixa` (Fator = 1)
    * **Cálculo:** `(15 * 20) / 1 = 300.0`
* **2. "Exportar e limpar dados de vendas" (Ana) - Pontuação: 300.0**
    * **Tempo:** 45 minutos
    * **Frequência:** `diária` (Fator = 20)
    * **Complexidade:** `Média` (Fator = 3)
    * **Cálculo:** `(45 * 20) / 3 = 300.0`
* **4. "Enviar lembrete de pendências no Slack" (Clara) - Pontuação: 200.0**
    * **Tempo:** 10 minutos
    * **Frequência:** `diária` (Fator = 20)
    * **Complexidade:** `Baixa` (Fator = 1)
    * **Cálculo:** `(10 * 20) / 1 = 200.0`

### Verificação do "Diagnóstico Estratégico"
* **"Dreno de Produtividade" para Ana (29,0 horas/mês):**
    Isso é calculado somando o tempo mensal total de todas as suas tarefas automatizáveis.
    * "Gerar relatório" (semanal): `60 min × 4 = 240 min`
    * "Exportar dados" (diária): `45 min × 20 = 900 min`
    * "Monitorar logs" (diária): `30 min × 20 = 600 min`
    * **Total:** `240 + 900 + 600 = 1740 minutos`
    * **Em Horas:** `1740 / 60 = 29,0 horas`

---

## Análise de Pontuação (Visão de Contribuição)
De um total de **1360 pontos** no pool de pontuação de automação em todas as tarefas, a priorização fica muito clara:
* **2 Principais Tarefas:** As tarefas de Beatriz ("Agendar reuniões") e Ana ("Exportar dados") cada uma contribui com **~22%** do total do pool de pontuação.
* **Impacto Combinado:** Juntas, essas duas tarefas diárias representam mais de **44%** de toda a oportunidade quantificável de automação, tornando-as as prioridades indiscutíveis.

---

## Diagnóstico Estratégico (Visão de Impacto)
A carga de trabalho total automatizável identificada para a equipe é de aproximadamente **43,3 horas por mês**.
* **Dreno de Produtividade de Ana:** As tarefas automatizáveis de Ana consomem **29,0 horas**, o que representa **~67%** de todo o dreno de produtividade identificado da equipe. Isso identifica seu fluxo de trabalho como a principal oportunidade de otimização.
* **Eficiência de "Quick Win":** Automatizar apenas as 3 principais tarefas libera **24,0 horas**. Esta ação capturaria **~55%** do total de economia de tempo possível com um esforço altamente focado.

### Análise de Pontuações Corrigidas
* **1. "Agendar reuniões de alinhamento" (Beatriz) - Pontuação: 300.0**
    * **Tempo:** 15 minutos
    * **Frequência:** `diária` (Fator = 20)
    * **Complexidade:** `Baixa` (Fator = 1)
    * **Cálculo:** `(15 * 20) / 1 = 300.0`
* **2. "Exportar e limpar dados de vendas" (Ana) - Pontuação: 300.0**
    * **Tempo:** 45 minutos
    * **Frequência:** `diária` (Fator = 20)
    * **Complexidade:** `Média` (Fator = 3)
    * **Cálculo:** `(45 * 20) / 3 = 300.0`
* **4. "Enviar lembrete de pendências no Slack" (Clara) - Pontuação: 200.0**
    * **Tempo:** 10 minutos
    * **Frequência:** `diária` (Fator = 20)
    * **Complexidade:** `Baixa` (Fator = 1)
    * **Cálculo:** `(10 * 20) / 1 = 200.0`

### Verificação do "Diagnóstico Estratégico"
* **"Dreno de Produtividade" para Ana (29,0 horas/mês):**
    Isso é calculado somando o tempo mensal total de todas as suas tarefas automatizáveis.
    * "Gerar relatório" (semanal): `60 min × 4 = 240 min`
    * "Exportar dados" (diária): `45 min × 20 = 900 min`
    * "Monitorar logs" (diária): `30 min × 20 = 600 min`
    * **Total:** `240 + 900 + 600 = 1740 minutos`
    * **Em Horas:** `1740 / 60 = 29,0 horas`

---

## Análise de Pontuação (Visão de Contribuição)
De um total de **1360 pontos** no pool de pontuação de automação em todas as tarefas, a priorização fica muito clara:
* **2 Principais Tarefas:** As tarefas de Beatriz ("Agendar reuniões") e Ana ("Exportar dados") cada uma contribui com **~22%** do total do pool de pontuação.
* **Impacto Combinado:** Juntas, essas duas tarefas diárias representam mais de **44%** de toda a oportunidade quantificável de automação, tornando-as as prioridades indiscutíveis.

---

## Diagnóstico Estratégico (Visão de Impacto)
A carga de trabalho total automatizável identificada para a equipe é de aproximadamente **43,3 horas por mês**.
* **Dreno de Produtividade de Ana:** As tarefas automatizáveis de Ana consomem **29,0 horas**, o que representa **~67%** de todo o dreno de produtividade identificado da equipe. Isso identifica seu fluxo de trabalho como a principal oportunidade de otimização.
* **Eficiência de "Quick Win":** Automatizar apenas as 3 principais tarefas libera **24,0 horas**. Esta ação capturaria **~55%** do total de economia de tempo possível com um esforço altamente focado.